### PR TITLE
fix: [CDS-51640]: add retry logic for workload watcher

### PR DESCRIPTION
**Pull request by achyuth-harness**
## Describe your changes
Going forward we will be adding the logic to retry the watch call in case of failures (similar to what we have in CG). We have also added SocketException to the list of exceptions which if we hit, we retry the call.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-ALL: `trigger tiAll`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/43892)

### CODE REVIEW

File: 960-api-services/src/main/java/io/harness/k8s/KubernetesContainerServiceImpl.java [hunk 1]
```diff
@@ -2288,10 +2297,44 @@ String generateKubeConfigStringForOpenID(KubernetesConfig config, OidcTokenReque
...
+          String errorMessage = "Failed to close Kubernetes watch." + ExceptionUtils.getMessage(ex);
+          log.error(errorMessage, ex);
+          return false;
+        } catch (ApiException e) {
+          ApiException ex = ExceptionMessageSanitizer.sanitizeException(e);
+          String errorMessage =
+              String.format("Failed to watch rollout status for workload [%s]. ", workload.kindNameRef())
+              + ExceptionUtils.getMessage(ex);
+          log.error(errorMessage, ex);
+          executionLogCallback.saveExecutionLog(errorMessage, LogLevel.ERROR);
```
Comment (abosii-harness) [inline 960-api-services/src/main/java/io/harness/k8s/KubernetesContainerServiceImpl.java:2322]: As discussed take care of proper exception handling. Currently this duplicate the logic in status watcher class. 
You can just always wrap it in throw new RuntimeException(e)

File: 960-api-services/src/main/java/io/harness/k8s/steadystate/watcher/workload/DaemonSetApiWatcher.java [hunk 2]
```diff
@@ -79,40 +75,30 @@ private boolean watchDaemonSet(ApiClient apiClient, KubernetesResourceId workloa
...
         return false;
-      } catch (RuntimeException e) {
-        if (e.getCause() instanceof InterruptedIOException) {
-          log.warn("Kubernetes watch was aborted.", e);
-          Thread.currentThread().interrupt();
-          return false;
-        }
-        log.error("Runtime exception during Kubernetes watch.", e);
+      }, false, executionLogCallback);
+    } catch (ApiException e) {
```
Comment (abosii-harness) [inline 960-api-services/src/main/java/io/harness/k8s/steadystate/watcher/workload/DaemonSetApiWatcher.java:83]: This logic can be moved to a common where you will pass additionally call and predicate so you wouldn't duplicate the code

**Review by abosii-harness — CHANGES_REQUESTED** [review]

File: 960-api-services/src/main/java/io/harness/k8s/KubernetesContainerServiceImpl.java [hunk 3]
```diff
@@ -2349,7 +2352,20 @@ public <T extends KubernetesObject> boolean watchWithRetries(
       }
       return false;
     });
-    return (v1Supplier == null || v1Supplier.get() == null) ? false : v1Supplier.get();
+    return v1Supplier.get();
+  }
+
+  private Type getK8sObjectType(String type) {
```
Comment (abosii-harness) [inline 960-api-services/src/main/java/io/harness/k8s/KubernetesContainerServiceImpl.java:2358]: why don't pass result of TypeToken<Watch.Response<V1Deployment>>() {}.getType() directly to the mehod?

**Review by abosii-harness — CHANGES_REQUESTED** [review]

**Review by achyuth-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43892)* status: 6 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`960-api-services/src/main/java/io/harness/k8s/KubernetesContainerServiceImpl.java` line 2322 at r2](https://reviewable.io/reviews/harness/harness-core/43892#-NOAMcTByh2wmmZKcJl9-r2-2322:-NOK6UT35h69IL-QQ7pv:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/1fd3e0a6ef563017a2baafc712a14a8211c67f32/960-api-services/src/main/java/io/harness/k8s/KubernetesContainerServiceImpl.java#L2322)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

As discussed take care of proper exception handling. Currently this duplicate the logic in status watcher class. 
You can just always wrap it in throw new RuntimeException(e)
</blockquote></details>

Done.
___
*[`960-api-services/src/main/java/io/harness/k8s/KubernetesContainerServiceImpl.java` line 2358 at r5](https://reviewable.io/reviews/harness/harness-core/43892#-NOAMcTByh2wmmZKcJl9-r5-2358:-NOK6VNI2shGOBhsVwcf:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/14956f6ee224c1ec4e035d023a7a0443fddacc4d/960-api-services/src/main/java/io/harness/k8s/KubernetesContainerServiceImpl.java#L2358)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

why don't pass result of TypeToken<Watch.Response<V1Deployment>>() {}.getType() directly to the mehod?
</blockquote></details>

Done.
___
*[`960-api-services/src/main/java/io/harness/k8s/steadystate/watcher/workload/DaemonSetApiWatcher.java` line 83 at r2](https://reviewable.io/reviews/harness/harness-core/43892#-NOAMcTByh2wmmZKcJlA-r2-83:-NOK6W8w1wBiq7Q-RUkF:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/1fd3e0a6ef563017a2baafc712a14a8211c67f32/960-api-services/src/main/java/io/harness/k8s/steadystate/watcher/workload/DaemonSetApiWatcher.java#L83)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

This logic can be moved to a common where you will pass additionally call and predicate so you wouldn't duplicate the code
</blockquote></details>

Done.

**Review by abhinavsingh1196 — CHANGES_REQUESTED** [review]
Please describe the change in the PR description.

Also, please test the following negative cases:
- for a long running steady state check (which then leads to step timeout), release history entry should be marked failed
- for a long running steady state check (which is aborted midway), release history entry should be marked failed
There should also not be any false positive exception logging (except the usual interrupts) since we are handling that case

> Reviewed 1 of 1 files at r3, 1 of 5 files at r4, 3 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43892#-:-NOOY6XB4c_E7hcAhc4F:bd75tbn)* status: 6 file(s) to review. 2 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`960-api-services/src/main/java/io/harness/k8s/KubernetesContainerServiceImpl.java` line 2323 at r6](https://reviewable.io/reviews/harness/harness-core/43892#-NOOXYX80c84u4KOPTJE:-NOOXYX9D3vgTJwTLuRg:beczx9j) ([raw file](https://github.com/harness/harness-core/blob/1aa3aac5aee6e3d4b9afb011f4e3c7fb4ef8f69e/960-api-services/src/main/java/io/harness/k8s/KubernetesContainerServiceImpl.java#L2323)):*
> ```Java
>       return false;
>     } catch (RuntimeException e) {
>       if (e.getCause() instanceof InterruptedIOException) {
> ```

add nullcheck for getCause() just to be safe
___
*[`960-api-services/src/main/java/io/harness/k8s/steadystate/watcher/workload/DaemonSetApiWatcher.java` line 50 at r6](https://reviewable.io/reviews/harness/harness-core/43892#-NOOVb6A8j3Tb0gYYa0g:-NOOVb6A8j3Tb0gYYa0h:b-p6vcne) ([raw file](https://github.com/harness/harness-core/blob/1aa3aac5aee6e3d4b9afb011f4e3c7fb4ef8f69e/960-api-services/src/main/java/io/harness/k8s/steadystate/watcher/workload/DaemonSetApiWatcher.java#L50)):*
> ```Java
>       LogCallback executionLogCallback, boolean errorFrameworkEnabled) throws Exception {
>     AppsV1Api appsV1Api = new AppsV1Api(apiClient);
>     return kubernetesContainerService.watchRetriesWrapper(new TypeToken<Watch.Response<V1DaemonSet>>() {}.getType(),
> ```

the third and fourth params can be moved to a separate method for better readability. Also, create a var for TypeToken

**Review by achyuth-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43892)* status: 7 file(s) to review. 2 change requests. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`960-api-services/src/main/java/io/harness/k8s/KubernetesContainerServiceImpl.java` line 2323 at r6](https://reviewable.io/reviews/harness/harness-core/43892#-NOOXYX80c84u4KOPTJE:-NOhmgN2DyExhCjHKBKd:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/1aa3aac5aee6e3d4b9afb011f4e3c7fb4ef8f69e/960-api-services/src/main/java/io/harness/k8s/KubernetesContainerServiceImpl.java#L2323)):*
<details><summary><i>Previously, abhinavsingh1196 (Abhinav Singh) wrote…</i></summary><blockquote>

add nullcheck for getCause() just to be safe
</blockquote></details>

Done.
___
*[`960-api-services/src/main/java/io/harness/k8s/steadystate/watcher/workload/DaemonSetApiWatcher.java` line 50 at r6](https://reviewable.io/reviews/harness/harness-core/43892#-NOOVb6A8j3Tb0gYYa0g:-NOhmhfN7apH6Y6XXgh0:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/1aa3aac5aee6e3d4b9afb011f4e3c7fb4ef8f69e/960-api-services/src/main/java/io/harness/k8s/steadystate/watcher/workload/DaemonSetApiWatcher.java#L50)):*
<details><summary><i>Previously, abhinavsingh1196 (Abhinav Singh) wrote…</i></summary><blockquote>

the third and fourth params can be moved to a separate method for better readability. Also, create a var for TypeToken 
</blockquote></details>

Done.

**Review by abosii-harness — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r3, 1 of 5 files at r4, 5 of 5 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43892)* status: 5 file(s) to review. 2 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`960-api-services/src/main/java/io/harness/k8s/KubernetesContainerServiceImpl.java` line 2332 at r7](https://reviewable.io/reviews/harness/harness-core/43892#-NOiRPk-5iB2wrWUaQSu:-NOiRPk-5iB2wrWUaQSv:b-o3inf1) ([raw file](https://github.com/harness/harness-core/blob/362edf05942290c7e4a11e85a0e89f58ff30f344/960-api-services/src/main/java/io/harness/k8s/KubernetesContainerServiceImpl.java#L2332)):*
> ```Java
>   }
> 
>   public boolean watchWithRetries(
> ```

It should be <T extends KubernetesObject> and then use Predicate<Watch.Response<T>> consumer (nit: not consumer but predicate)
___
*[`960-api-services/src/main/java/io/harness/k8s/steadystate/watcher/workload/DaemonSetApiWatcher.java` line 50 at r6](https://reviewable.io/reviews/harness/harness-core/43892#-NOOVb6A8j3Tb0gYYa0g:-NOiQxsJEJRRZjhZct4p:b-o4u7zm) ([raw file](https://github.com/harness/harness-core/blob/1aa3aac5aee6e3d4b9afb011f4e3c7fb4ef8f69e/960-api-services/src/main/java/io/harness/k8s/steadystate/watcher/workload/DaemonSetApiWatcher.java#L50)):*
<details><summary><i>Previously, achyuth-harness (Achyuth Prakash) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

nit: workloadDetails and supplier can be moved out to local variable 
ex: workloadDetails = new WorkloadDetails(v1DaemonSetType, apiClient, workload, executionLogCallback, errorFrameworkEnabled)
callSupplier = () -> appsV1Api.listNamespacedDaemonSetCall(workload.getNamespace(), null, null, null, null, null, null, null,
                null, WATCH_CALL_TIMEOUT_SECONDS, true, null),
___
*[`960-api-services/src/main/java/io/harness/k8s/steadystate/watcher/workload/DaemonSetApiWatcher.java` line 59 at r7](https://reviewable.io/reviews/harness/harness-core/43892#-NOiRGJ-43kgiSDydC18:-NOiRGJ-43kgiSDydC19:bn8v1qa) ([raw file](https://github.com/harness/harness-core/blob/362edf05942290c7e4a11e85a0e89f58ff30f344/960-api-services/src/main/java/io/harness/k8s/steadystate/watcher/workload/DaemonSetApiWatcher.java#L59)):*
> ```Java
>                 null, WATCH_CALL_TIMEOUT_SECONDS, true, null),
>         event -> {
>           V1DaemonSet daemonSet = (V1DaemonSet) event.object;
> ```

This cast shouldn't be required, this should be resolvable at compile if your method will be templatized

**Review by achyuth-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43892)* status: 5 file(s) to review. 2 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`960-api-services/src/main/java/io/harness/k8s/KubernetesContainerServiceImpl.java` line 2332 at r7](https://reviewable.io/reviews/harness/harness-core/43892#-NOiRPk-5iB2wrWUaQSu:-NOjE7Rf4gd6twf6opMG:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/362edf05942290c7e4a11e85a0e89f58ff30f344/960-api-services/src/main/java/io/harness/k8s/KubernetesContainerServiceImpl.java#L2332)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

It should be <T extends KubernetesObject> and then use Predicate<Watch.Response<T>> consumer (nit: not consumer but predicate)
</blockquote></details>

Done.
___
*[`960-api-services/src/main/java/io/harness/k8s/steadystate/watcher/workload/DaemonSetApiWatcher.java` line 50 at r6](https://reviewable.io/reviews/harness/harness-core/43892#-NOOVb6A8j3Tb0gYYa0g:-NOjE8YYBDfqZpSz--PI:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/1aa3aac5aee6e3d4b9afb011f4e3c7fb4ef8f69e/960-api-services/src/main/java/io/harness/k8s/steadystate/watcher/workload/DaemonSetApiWatcher.java#L50)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

nit: workloadDetails and supplier can be moved out to local variable 
ex: workloadDetails = new WorkloadDetails(v1DaemonSetType, apiClient, workload, executionLogCallback, errorFrameworkEnabled)
callSupplier = () -> appsV1Api.listNamespacedDaemonSetCall(workload.getNamespace(), null, null, null, null, null, null, null,
                null, WATCH_CALL_TIMEOUT_SECONDS, true, null),

</blockquote></details>

Done.
___
*[`960-api-services/src/main/java/io/harness/k8s/steadystate/watcher/workload/DaemonSetApiWatcher.java` line 59 at r7](https://reviewable.io/reviews/harness/harness-core/43892#-NOiRGJ-43kgiSDydC18:-NOjE9Hu1kKTix_Ol3kJ:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/362edf05942290c7e4a11e85a0e89f58ff30f344/960-api-services/src/main/java/io/harness/k8s/steadystate/watcher/workload/DaemonSetApiWatcher.java#L59)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

This cast shouldn't be required, this should be resolvable at compile if your method will be templatized
</blockquote></details>

Done.

**Review by abosii-harness — APPROVED** [review]
> Reviewed 4 of 4 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43892)* status: 5 file(s) to review. 1 change requests. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by abhinavsingh1196 — APPROVED** [review]
> Reviewed 1 of 5 files at r7, 4 of 4 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43892)* status: :shipit: complete! ✓
